test/integration/h2_fuzz.cc

```diff
@@ -0,0 +1,188 @@
+    uint32_t stream_idx = proto_frame.malformed_request_with_zerolen_header().stream_index();
+    absl::string_view host = proto_frame.malformed_request_with_zerolen_header().host();
+    absl::string_view path = proto_frame.malformed_request_with_zerolen_header().path();
+    ENVOY_LOG_MISC(trace, "Sending malformed_request_with_zerolen_header");
+    h2_frame = Http2Frame::makeMalformedRequestWithZerolenHeader(stream_idx, host, path);
+  } break;
+  case test::integration::H2TestFrame::kRequest: {
+    uint32_t stream_idx = proto_frame.request().stream_index();
+    absl::string_view host = proto_frame.request().host();
+    absl::string_view path = proto_frame.request().path();
+    ENVOY_LOG_MISC(trace, "Sending request");
+    h2_frame = Http2Frame::makeRequest(stream_idx, host, path);
+  } break;
+  case test::integration::H2TestFrame::kPostRequest: {
+    uint32_t stream_idx = proto_frame.post_request().stream_index();
+    absl::string_view host = proto_frame.post_request().host();
+    absl::string_view path = proto_frame.post_request().path();
+    ENVOY_LOG_MISC(trace, "Sending post request");
+    h2_frame = Http2Frame::makePostRequest(stream_idx, host, path);
+  } break;
+  default:
+    ENVOY_LOG_MISC(debug, "Proto-frame not supported!");
+    break;
+  }
+
+  write_func(h2_frame);
+}
+
+void H2FuzzIntegrationTest::replay(const test::integration::H2CaptureFuzzTestCase& input) {
+  initialize();
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("http"));
+  FakeRawConnectionPtr fake_upstream_connection;
+  bool stop_further_inputs = false;
+  for (int i = 0; i < input.events().size(); ++i) {
+    if (stop_further_inputs) {
+      break;
+    }
+    const auto& event = input.events(i);
+    ENVOY_LOG_MISC(debug, "Processing event: {}", event.DebugString());
+    // If we're disconnected, we fail out.
+    if (!tcp_client->connected()) {
+      ENVOY_LOG_MISC(debug, "Disconnected, no further event processing.");
+      break;
+    }
+    switch (event.event_selector_case()) {
+    // TODO(adip): Add ability to test complex interactions when receiving data
+    case test::integration::Event::kDownstreamSendEvent: {
+      auto downstream_write_func = [&](const Http2Frame& h2_frame) -> void {
+        tcp_client->write(std::string(h2_frame), false, false);
+      };
+      // Start H2 session - send hello string
+      tcp_client->write(Http2Frame::Preamble, false, false);
+      for (auto& frame : event.downstream_send_event().h2_frames()) {
+        if (!tcp_client->connected()) {
+          ENVOY_LOG_MISC(debug,
+                         "Disconnected, avoiding sending data, no further event processing.");
+          break;
+        }
+
+        ENVOY_LOG_MISC(trace, "sending downstream frame");
+        sendFrame(frame, downstream_write_func);
+      }
+    } break;
+    case test::integration::Event::kUpstreamSendEvent:
```

Comment:
Shouldn't need any of this stuff if in direct response mode.

Reply:
I think in direct response mode this actually does the wrong thing on second consideration, since we're replaying upstream events that should never occur.

Reply:
Ping on this one; do we want to just have direct response mode? Do you know if we have any coverage win from the full e2e in practice?

Reply:
I've left both of them. The non direct-response mode seems to be testing a bit more (as expected), but may end up spending much more time on the non-HCM/codec/router code that we aim to fuzz.
By keeping them both we can decide whether both should run or not.